### PR TITLE
ARROW-16772: [C++] Implement encode and decode functions for Run-Length encoding

### PR DESCRIPTION
**Pull request by zagto**

### CODE REVIEW

**Comment by westonpace** [comment]
@drin do you want to take a look through this PR

File: cpp/src/arrow/util/rle_util.cc [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                             accumulated_run_lengths + physical_length, logical_offset);
+  return std::distance(accumulated_run_lengths, it);
+}
+
+int64_t FindPhysicalIndex(const int32_t* accumulated_run_lengths,
+                          int64_t physical_length, int64_t physical_offset, int64_t logical_offset, int64_t logical_index) {
+  auto it = std::upper_bound(accumulated_run_lengths + physical_offset,
+                             accumulated_run_lengths + physical_offset + physical_length,
+                             logical_index + logical_offset);
+  return std::distance(accumulated_run_lengths, it);
```
Comment (drin) [inline cpp/src/arrow/util/rle_util.cc:20]: should this accommodate `physical_offset`?

File: cpp/src/arrow/util/rle_util_test.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  ASSERT_EQ(FindPhysicalOffset((const int32_t[]){2, 3, 4}, 3, 0), 0);
+  ASSERT_EQ(FindPhysicalOffset((const int32_t[]){2, 3, 4}, 3, 1), 0);
+  ASSERT_EQ(FindPhysicalOffset((const int32_t[]){2, 3, 4}, 3, 2), 1);
+  ASSERT_EQ(FindPhysicalOffset((const int32_t[]){2, 3, 4}, 3, 3), 2);
+  ASSERT_EQ(FindPhysicalOffset((const int32_t[]){2, 4, 6}, 3, 3), 1);
+}
+
+TEST(TestRleUtil, FindPhysicalIndexTest) {
+  ASSERT_EQ(FindPhysicalIndex((const int32_t[]){1, 10, 20, 30, 40}, 5, 2, 15, 14), 3);
+  ASSERT_EQ(FindPhysicalIndex((const int32_t[]){1, 10, 20, 30, 40}, 5, 2, 15, 15), 4);
```
Comment (drin) [inline cpp/src/arrow/util/rle_util_test.cc:44]: it occurs to me that the expected values here are relative to the whole RLE list, rather than relative to the physical offset provided. I am not sure if that is intended or not?

**Review by drin — COMMENTED** [review]
reviewed most of the files.

still need to go over:
* `vector_run_length_encode_test.cc`
* and `vector_run_length_encode.cc`

(not sure how the review interface works, so this is also a note for myself

File: cpp/src/arrow/compute/kernels/vector_run_length_encode_test.cc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    result.input = builder.Finish().ValueOrDie();
+    result.expected_values = result.input;
+    result.expected_run_lengths = {1, 2, 3};
+    result.string = "Type min, max, & null values";
+    return result;
+  }
+
+  std::shared_ptr<Array> input;
+  std::shared_ptr<Array> expected_values;
+  std::vector<int32_t> expected_run_lengths;
```
[the remark this answers is not in the document]
Comment (zagto) [reply cpp/src/arrow/compute/kernels/vector_run_length_encode_test.cc:58]: You're right, run_lengths is not an accurate name for this. I have started using the name run_ends for this in some places, and should probably here too

[the remark this answers is not in the document]
Comment (drin) [reply cpp/src/arrow/compute/kernels/vector_run_length_encode_test.cc:58]: `run_ends` is a really good name, IMO. So, that sounds great

**Comment by zagto** [comment]
@drin I hope this is not too confusing, but I did a lot of work on the rle-util branch in the recent time, which was not merged/pushed here yet, sorry for that. I'm going to push that, so that you can see the lastest version of the encode/decode kernels

Thanks for the feedback so far, I think the majority still applies to the current code. I don't think you have to review all the new rle_util code unless you want to, someone is eventually going to review [ARROW-17263](https://issues.apache.org/jira/browse/ARROW-17263) for that.

File: cpp/src/arrow/compute/api_vector.h [hunk 4]
```diff
@@ -580,6 +587,26 @@ Result<Datum> DictionaryEncode(
...
 
+/// \brief Run-Length-encode values in an array-like object
+/// \param[in] value array-like input
+/// \param[in] ctx the function execution context, optional
+/// \return result with same shape and type as input
+///
+/// \since 9.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> RunLengthEncode(const Datum& value, ExecContext* ctx = NULLPTR);
```
[the remark this answers is not in the document]
Comment (zagto) [reply cpp/src/arrow/compute/api_vector.h:598]: Yes, there should be once there are options for this function. I should probably remove RunLengthEncodeOptions for now, it does not have any options

File: cpp/src/arrow/util/rle_util.cc (hunk 1 above)
Comment (zagto) [reply cpp/src/arrow/util/rle_util.cc]: Yes! I ended up moving that logic to a function called GetPhysicalLength, since that is what it ended up being used for, but did not notice the wrong end iterator to upper_bound here

File: cpp/src/arrow/testing/gtest_util.cc [hunk 5]
```diff
@@ -101,8 +101,10 @@ std::vector<Type::type> AllTypeIds() {
           Type::DENSE_UNION,
           Type::SPARSE_UNION,
           Type::DICTIONARY,
+          Type::RUN_LENGTH_ENCODED,
```
Comment (drin) [inline cpp/src/arrow/testing/gtest_util.cc:104]: I think this type is duplicated on line 107, so probably only need one.

Comment (drin) [reply cpp/src/arrow/testing/gtest_util.cc:104]:
```suggestion
```

File: cpp/src/arrow/visitor_generate.h [hunk 6]
```diff
@@ -35,34 +35,39 @@ namespace arrow {
...
-  ACTION(List);                                 \
-  ACTION(LargeList);                            \
-  ACTION(Map);                                  \
-  ACTION(FixedSizeList);                        \
-  ACTION(Struct);                               \
-  ACTION(SparseUnion);                          \
-  ACTION(DenseUnion);                           \
-  ACTION(Dictionary);                           \
+// all types that can exist as a Scalar
+#define ARROW_GENERATE_FOR_ALL_SCALAR_TYPES(ACTION) \
```
Comment (drin) [inline cpp/src/arrow/visitor_generate.h:39]: Can all of these be scalars? If so, I'm not sure why `ARROW_GENERATE_FOR_ALL_TYPES` is being changed to a wrapper around this macro.

Otherwise, maybe a comment for `ARROW_GENERATE_FOR_ALL_TYPES` could help with clarity.

**Comment by drin** [comment]
> @drin ...I did a lot of work on the rle-util branch in the recent time... I'm going to push that, so that you can see the lastest version of the encode/decode kernels
> 
> Thanks for the feedback so far, I think the majority still applies to the current code. I don't think you have to review all the new rle_util code unless you want to, someone is eventually going to review [ARROW-17263](https://issues.apache.org/jira/browse/ARROW-17263) for that.

just saw this comment, thanks for the heads up! I'll try to review anyways, it hopefully doesn't have too much extra learning curve

File: cpp/src/arrow/array/array_encoded.cc [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+RunLengthEncodedArray::RunLengthEncodedArray(const std::shared_ptr<DataType>& type,
+                                             int64_t length,
+                                             const std::shared_ptr<Array>& run_ends_array,
+                                             const std::shared_ptr<Array>& values_array,
+                                             int64_t offset) {
+  ARROW_CHECK_EQ(type->id(), Type::RUN_LENGTH_ENCODED);
+  SetData(ArrayData::Make(type, length, {NULLPTR}, 0, offset));
+  data_->child_data.push_back(std::move(run_ends_array->data()));
+  data_->child_data.push_back(std::move(values_array->data()));
```
Comment (drin) [inline cpp/src/arrow/array/array_encoded.cc:41]: Is there design documentation somewhere?

I think this approach makes sense, but I'm just wondering what the trade-offs are compared to having run ends and values in a single Array, in a similar layout as [variable-sized binary](https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout)? I suspect this approach is just more flexible, especially for run length encoding a nested type array; though, there is likely more to it than that.

File: cpp/src/arrow/visitor_generate.h (hunk 6 above)
Comment (zagto) [reply cpp/src/arrow/visitor_generate.h:39]: I think all types that exist in arrow right now can exist as scalars. I was considering adding an RLE scalar class for consistency. But the encoding can only be applied to arrays, so an RLE scalar would just be a wrapper for another scalar of the actual data type. To me that seems like something you would not actually want to use and therefore it should not exist.

The idea for new macro is code that works on scalar values does not need to generate a case for RLE (and possibly other future array encodings), which would be unnecessary and may also result in using `type_traits<RunLengthEncoded>::ScalarType`

File: cpp/src/arrow/array/array_encoded.cc (hunk 7 above)
Comment (zagto) [reply cpp/src/arrow/array/array_encoded.cc:41]: I don't think I have written that down anywhere, so I'll put it here:

Yes having the values buffer(s) in the parent would makes supporting multiple data types more complicated:
- Each type in arrow has a defined buffer layout, RLE would not. It would have different buffers depending on the type it encodes, i.e. a single values buffer for integer types, typeID+offset for dense union..., but also add a run ends buffer somewhere
- we could not rely on things like Validate(), unary kernels, array builders, or anything else to work on the values array without modification
- While in Arrow C++ the parent array has pointer to the exact DataType object, in the format itself an array has only a type ID, RLE in this case. The type of the values would have to be put into the buffers somewhere, making the layout even more complicated

Having the run ends as a buffer in the parent is less bad, but also causes problems:
- When a buffer is passed over the C bridge, we don't really know it's size (or at least how I understand it). It can be determined either from the length field, or a combination of that and other buffers easily for other types, but not for RLE. The RLE format is designed to allow O(log(n)) random access ,using binary search. For Binary search, we need an upper end, which we conveniently get from the length field of a child array (Yes the whole array needs to be sorted).
- Also having a child array gives us another type field, which means the format could support RLE arrays with  run ends that are not int32 in the future.

In general I like the idea that a user can simply run any analysis on the run ends and values array on thier own, if they find it useful for some reason. They could always construct an array from the buffers but it just seems less messy with separate arrays.

Comment (drin) [reply cpp/src/arrow/array/array_encoded.cc:41]: okay, I think I figured it out, but just to confirm: this approach allows you to keep binary data contiguous (which you need offsets for), which prevents you from reasonably holding run ends and run values in a single ArrayData?

**Review by drin — COMMENTED** [review]
I checked out most of `vector_run_length_encode.cc`. I think I understand the memory layout now, so hopefully finishing the rest of the review goes quick!

All my suggestions are minor, great job so far 🎉 !

File: cpp/src/arrow/array/array_encoded.cc (hunk 7 above)
Comment (drin) [reply cpp/src/arrow/array/array_encoded.cc:41]: awesome, thanks! I figured out the memory layout as well, so I realized that this approach also allows run values to be a binary type with a contiguous buffer.

File: cpp/src/arrow/compute/kernels/vector_run_length_encode.cc [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+    values_array_data->buffers.push_back(std::move(values_buffer));
+    run_ends_array_data->buffers.push_back(NULLPTR);
+    run_ends_array_data->buffers.push_back(std::move(run_lengths_buffer));
+
+    output_array_data->null_count.store(input_null_count);
+    values_array_data->null_count = output_null_count;
+
+    this->output_validity = values_array_data->template GetMutableValues<uint8_t>(0);
+    this->output_values = values_array_data->template GetMutableValues<uint8_t>(1);
+    auto output_run_ends = run_ends_array_data->template GetMutableValues<int32_t>(1);
```
Comment (drin) [inline cpp/src/arrow/compute/kernels/vector_run_length_encode.cc:184]:
```suggestion
    // set mutable pointers for output; `WriteValue` uses member `output_` variables
    this->output_validity = values_array_data->template GetMutableValues<uint8_t>(0);
    this->output_values = values_array_data->template GetMutableValues<uint8_t>(1);
    auto output_run_ends = run_ends_array_data->template GetMutableValues<int32_t>(1);
```

I think it would be helpful to add a comment so that it's clearer to a reader that these member variables are aliases and are used by WriteValue. It can be hard to follow the logic otherwise.

File: cpp/src/arrow/array/array_encoded_test.cc [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+  ASSERT_OK_AND_ASSIGN(rle_array,
+                       RunLengthEncodedArray::Make(int32_values, int32_values, 3));
+  ASSERT_EQ(rle_array->length(), 3);
+  ASSERT_ARRAYS_EQUAL(*rle_array->values_array(), *int32_values);
+  ASSERT_ARRAYS_EQUAL(*rle_array->run_ends_array(), *int32_values);
+  ASSERT_EQ(rle_array->offset(), 0);
+  ASSERT_EQ(rle_array->data()->null_count, 0);
+  // one dummy buffer, since code may assume there is exactly one buffer
+  ASSERT_EQ(rle_array->data()->buffers.size(), 1);
```
[the remark this answers is not in the document]
Comment (zagto) [reply cpp/src/arrow/array/array_encoded_test.cc:69]: This is correct since the RLE parent always has to have a null count of 0, and no validity bitmap. It can actually encode arrays with null values, but there are not set in the parent, since we don't want the standard on bit per logical element validity bitmap. I find this a bit confusing, but this is how it works for Unions too. I tried to do it differently first, but Arrow code essentially assumes non-zero null count = has validity bitmap

File: cpp/src/arrow/compute/kernels/vector_run_length_encode.cc [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+  using typename EncodeDecodeCommonExec<ArrowType, has_validity_buffer>::CType;
+  using typename EncodeDecodeCommonExec<ArrowType, has_validity_buffer>::Element;
+
+  Status Exec() {
+    ArrayData* output_array_data = this->exec_result->array_data().get();
+    if (this->input_array.length == 0) {
+      output_array_data->length = 0;
+      output_array_data->offset = 0;
+      output_array_data->buffers = {NULLPTR};
+      output_array_data->child_data[0] =
```
Comment (zagto) [inline cpp/src/arrow/compute/kernels/vector_run_length_encode.cc:111]: note to self: this doesn't look right
